openmls/src/group/public_group/mod.rs

```diff
@@ -0,0 +1,265 @@
+                let removed_index = remove_proposal.removed();
+                if removed_index < free_leaf_index {
+                    removed_index
+                } else {
+                    free_leaf_index
+                }
+            } else {
+                return Err(LibraryError::custom("missing key package"));
+            }
+        } else {
+            free_leaf_index
+        };
+        Ok(leaf_index)
+    }
+
+    /// Create an empty  [`PublicGroupDiff`] based on this [`PublicGroup`].
+    pub(crate) fn empty_diff(&self) -> PublicGroupDiff {
+        PublicGroupDiff::new(self)
+    }
+
+    /// Merge the changes performed on the [`PublicGroupDiff`] into this
+    /// [`PublicGroup`].
+    pub(crate) fn merge_diff(&mut self, diff: StagedPublicGroupDiff) {
+        self.treesync.merge_diff(diff.staged_diff);
+        self.group_context = diff.group_context;
+        self.interim_transcript_hash = diff.interim_transcript_hash;
+        self.confirmation_tag = diff.confirmation_tag;
+    }
+
+    /// Derives [`EncryptionKeyPair`]s for the nodes in the shared direct path
+    /// of the leaves with index `leaf_index` and `sender_index`.  This function
+    /// also checks that the derived public keys match the existing public keys.
+    ///
+    /// Returns the [`CommitSecret`] derived from the path secret of the root
+    /// node, as well as the derived [`EncryptionKeyPair`]s. Returns an error if
+    /// the target leaf is outside of the tree.
+    ///
+    /// Returns [`DerivePathError::PublicKeyMismatch`] if the derived keys don't
+    /// match with the existing ones.
+    ///
+    /// Returns [`DerivePathError::LibraryError`] if the sender_index is not
+    /// in the tree.
+    pub(crate) fn derive_path_secrets(
+        &self,
+        backend: &impl OpenMlsCryptoProvider,
+        ciphersuite: Ciphersuite,
+        path_secret: PathSecret,
+        sender_index: LeafNodeIndex,
+        leaf_index: LeafNodeIndex,
+    ) -> Result<(Vec<EncryptionKeyPair>, CommitSecret), DerivePathError> {
+        self.treesync.derive_path_secrets(
+            backend,
+            ciphersuite,
+            path_secret,
+            sender_index,
+            leaf_index,
+        )
+    }
+}
+
+// Getters
+impl PublicGroup {
+    pub fn ciphersuite(&self) -> Ciphersuite {
+        self.group_context.ciphersuite()
+    }
+
+    pub fn version(&self) -> ProtocolVersion {
+        self.group_context.protocol_version()
+    }
+
+    pub fn group_id(&self) -> &GroupId {
+        self.group_context.group_id()
+    }
+
+    pub fn group_context(&self) -> &GroupContext {
+        &self.group_context
+    }
+
+    pub fn extensions(&self) -> &Extensions {
+        self.group_context.extensions()
+    }
+
+    pub fn required_capabilities(&self) -> Option<&RequiredCapabilitiesExtension> {
+        self.group_context.required_capabilities()
+    }
+
+    pub(crate) fn treesync(&self) -> &TreeSync {
```

Comment:
I assume that this is supposed to go away in follow ups?

Reply:
Exactly. I thought about doing it in this one, but treesync is used externally in quite a few places, so I decided on punting it.